api/src/main/java/org/apache/iceberg/actions/ConfigurableAction.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.actions;
+
+import java.util.Map;
+
+/**
+ * An action performed on a table that accepts options that control the execution.
+ *
+ * @param <R> the Java type of the result produced by this action
+ */
+public interface ConfigurableAction<ThisT, R> extends Action<R> {
+  /**
+   * Configures this action with an extra option.
+   *
+   * @param name an option name
+   * @param value an option value
+   * @return this for method chaining
+   */
+  ThisT withOption(String name, String value);
```

Comment:
There will be configs specific to engines. For example, whether to use `toLocalIterator` in Spark or not. It does not make sense to expose such a method in this API as other engines may not support it.

There are two approaches we can take:
- Support a map of options in the API.
- Have engine-specific interfaces that extend base action interfaces. For example, `ExpireSnapshotsSparkAction` will extend `ExpireSnapshotsAction`.

I went with the first option as it seems simpler. I can reconsider depending on what the community thinks.

Reply:
I think this follows some of our discussion yesterday. I would prefer that any options that a user would not modify in 90% of use cases be an "option", I think engine specific things fall into that category. The example of `toLocalIterator` is a good one since it takes a much more advanced Spark user to understand the requirement so we should hide that in the public api exposed via any java IDE.

Reply:
I don't love an opque map here but I see the purpose.

Reply:
Sounds like an intermediate consensus then. I'll keep it as is for now.

Reply:
I think we could move this to the default Action interface if we wanted to make it a bit simpler.

Reply:
Do you mean `withOptions` or engine-specific methods, @rdblue?

Reply:
I'll resolve this one as we agreed in another thread.